### PR TITLE
Unification of fe fields

### PR DESCRIPTION
**Pull request by vlukes**
FE fields `VolumeField` and `SurfaceField` are unified in `FEField`.

### CODE REVIEW

File: sfepy/discrete/fem/fields_base.py [hunk 1]
```diff
@@ -236,10 +242,6 @@ def __init__(self, name, dtype, shape, region, approx_order=1):
         Assumes one cell type for the whole region!
         """
         shape = parse_shape(shape, region.domain.shape.dim)
-        if not self._check_region(region):
-            raise ValueError('unsuitable region for field %s! (%s)' %
-                             (name, region.name))
-
         Struct.__init__(self, name=name, dtype=dtype, shape=shape,
```
Comment (rc) [inline sfepy/discrete/fem/fields_base.py:245]: Does it mean all regions are now usable? Or just cell and facet ones?

File: sfepy/discrete/fem/fields_base.py [hunk 2]
```diff
@@ -629,7 +702,8 @@ def restore_substituted(self, vec):
         vec = vec.reshape((self.n_nod, self.n_components)).copy()
         evec = vec[self.econn]
 
-        vec[self.econn0] = nm.einsum('cji,cjk->cik', self.basis_transform, evec)
+        vec[self.econn0] = nm.einsum('cji,cjk->cik', self.basis_transform,
+                                     evec)
```
Comment (rc) [inline sfepy/discrete/fem/fields_base.py:706]: Maybe turn on optimization too? But for two operands it IMO does not matter.

**Comment by rc** [comment]
Thanks!

I noticed that the volume->cell, surface->facet change needs to be done also in the docs.

File: sfepy/discrete/fem/fields_base.py [hunk 3]
```diff
@@ -1438,42 +1257,53 @@ def average_qp_to_vertices(self, data_qp, integral):
...
 
         return data_vertex
 
+    # from SurfaceField()
+    # def _setup_bubble_dofs(self):
+    #     """
+    #     Setup bubble DOF connectivity.
+    #     """
+    #     return 0, None, None
+
```
Comment (rc) [inline sfepy/discrete/fem/fields_base.py:1309]: SurfaceField was defined on faces of a volume mesh, so it did not have bubble DOFs, only vertex, edge and face. How is this solved now?

File: sfepy/discrete/fem/fields_base.py (hunk 2 above)
Comment (vlukes) [reply sfepy/discrete/fem/fields_base.py]: `optimization=True` costs nothing and does no harm

File: sfepy/discrete/fem/fields_base.py (hunk 1 above)
Comment (vlukes) [reply sfepy/discrete/fem/fields_base.py:245]: Yes, just cell and facet regions are usable. Fixed in `._setup_geometry()`.

File: sfepy/discrete/fem/fields_base.py (hunk 3 above)
Comment (vlukes) [reply sfepy/discrete/fem/fields_base.py:1309]: What should return this function for `VolumeField` (`self.is_surface` is `False`)?

Comment (rc) [reply sfepy/discrete/fem/fields_base.py:1309]: See e.g. https://github.com/sfepy/sfepy/blob/master/sfepy/discrete/fem/fields_nodal.py

**Comment by rc** [comment]
LGTM from my side, thanks. Let me know when it is ready. You plan to update the docs, right?

**Comment by rc** [comment]
I tried to update the docs, please check when you have time.

**Comment by rc** [comment]
So I guess it is done now. I will merge soon.

**Comment by vlukes** [comment]
> So I guess it is done now. I will merge soon.

Do it.

**Pull request merged**